// @mysten/deepbook-v3/dist/contracts/utils/index.mjs
import { BcsStruct, TypeTagSerializer, bcs } from "@mysten/sui/bcs";
import "@mysten/sui/transactions";
import { normalizeStructTag, normalizeSuiAddress } from "@mysten/sui/utils";

//#region src/contracts/utils/index.ts
const MOVE_STDLIB_ADDRESS = normalizeSuiAddress("0x1");
const SUI_FRAMEWORK_ADDRESS = normalizeSuiAddress("0x2");
const HAS_PHANTOM_REGEX = /phantom [A-Za-z_$][A-Za-z0-9_$]*/;
function splitTopLevelTypeArgs(inner) {
	const parts = [];
	let depth = 0;
	let current = "";
	for (const char of inner) {
		if (char === "," && depth === 0) {
			parts.push(current.trim());
			current = "";
			continue;
		}
		if (char === "<") depth++;
		if (char === ">") depth--;
		current += char;
	}
	if (current) parts.push(current.trim());
	return parts;
}
function buildTypeTag(name, options) {
	const lt = name.indexOf("<");
	const base = lt === -1 ? name : name.slice(0, lt);
	if (base.split("::").length !== 3) throw new Error(`${name} is not a top-level Move type`);
	let result = name;
	if (options?.typeArguments) {
		const baked = lt === -1 ? [] : splitTopLevelTypeArgs(name.slice(lt + 1, -1));
		const supplied = options.typeArguments.map((arg) => {
			if (typeof arg === "string") return arg;
			if (arg && typeof arg.serialize === "function" && typeof arg.name === "string") return arg.name;
			throw new Error(`Invalid type argument ${stringify(arg)}`);
		});
		if (supplied.length !== baked.length) throw new Error(`Expected ${baked.length} type arguments for ${base}, got ${supplied.length}`);
		result = supplied.length === 0 ? base : `${base}<${supplied.join(", ")}>`;
	}
	if (HAS_PHANTOM_REGEX.test(result)) throw new Error(options?.typeArguments ? `A type argument contains an unfilled phantom parameter in ${result}` : `Missing type arguments for ${result}`);
	if (options?.package) {
		const [, ...rest] = result.split("::");
		result = [options.package, ...rest].join("::");
	}
	if (!HAS_PHANTOM_REGEX.test(result) && !/[@/]/.test(result)) TypeTagSerializer.parseFromStr(result);
	return result;
}
async function resolveBuiltTypeTag(name, options) {
	const { client, ...rest } = options;
	const { type } = await client.core.mvr.resolveType({ type: buildTypeTag(name, rest) });
	return normalizeStructTag(type);
}
var MoveStruct = class extends BcsStruct {
	/**
	* Build the type tag for this struct.
	*
	* `typeArguments` is the full positional list, in Move declaration order, and
	* is required when the struct has unfilled phantom parameters. The result may
	* contain MVR names: those are valid in transaction `typeArguments`, but for
	* queries or comparisons against on-chain data use `resolveTypeTag` instead.
	*/
	typeTag(...args) {
		return buildTypeTag(this.name, args[0]);
	}
	/**
	* Build the type tag for this struct, then resolve any MVR names through the
	* client (using its configured overrides and the MVR API) and return the
	* normalized, address-only form suitable for queries and comparisons against
	* on-chain data.
	*/
	async resolveTypeTag(options) {
		return resolveBuiltTypeTag(this.name, options);
	}
	async get({ objectId, ...options }) {
		const [res] = await this.getMany({
			...options,
			objectIds: [objectId]
		});
		if (!res) throw new Error(`No object found for id ${objectId}`);
		return res;
	}
	async getMany({ client, ...options }) {
		return (await client.core.getObjects({
			...options,
			include: {
				...options.include,
				content: true
			}
		})).objects.map((obj) => {
			if (obj instanceof Error) throw obj;
			return {
				...obj,
				json: this.parse(obj.content)
			};
		});
	}
};
function stringify(val) {
	if (typeof val === "object") return JSON.stringify(val, (_key, value) => typeof value === "bigint" ? value.toString() : value);
	if (typeof val === "bigint") return val.toString();
	return val;
}

//#endregion
export { MoveStruct };
